Advanced mapping

kpob opened the pull request:
Resolves #38 
* `Mapping` and `Variable` serialization
* Add `get_instance()` function to `Mapping`
* Add `map` macro to simplify complex mappings syntax, eg. `map!(mapping[key1][key2][key3])`  
* contract state stored under a single key
* update contact initialization - add a named key `state`

kpob committed:
A module can be a mapping value

kpob committed:
Handle nested mappings

kpob committed:
fix clippy warning

kpob committed:
More complex example
* a value of Mapping is made of modules

kpob committed:
Fix serialization

kpob committed:
fmt

kpob committed:
Update odra-ci.yml

kpob committed:
simplify nested mapping init

kpob committed:
use a single dict for the state

kpob committed:
add map macro

kpob committed:
Fix contract setup

kpob committed:
Add get_instance function

kpob committed:
refactor map macro

kpob committed:
pass attributes to Ref's functions

kpob committed:
update Mapping, fix examples

kpob committed:
Merge branch 'release/0.2.1' into feature/advanced_mapping

kpob committed:
fix

kpob committed:
fix clippy warning

kpob committed:
fmt

kpob committed:
Fix test

kpob committed:
refactor example

zie1ony commented on core/src/mapping.rs:
```diff
@@ -41,6 +46,25 @@ impl<K: OdraType + Hash, V: OdraType + Default> Mapping<K, V> {
+            V::instance(&namespace)
+        }
+
+        #[cfg(feature = "casper")]
```
There is no need to hex namespace here. Hashing should only happen right after reading from the storage.

kpob replied:
There is no namespace hashing here. We need to build the path but the `key`  is not a string, even is not strigifiable, so the only thing that came to my mind to have string representation of the key, was to serialize the key and use it's hex representation.

zie1ony commented on core/src/mapping.rs:
```diff
@@ -79,6 +103,58 @@ impl<K: OdraType + Hash, V: OdraType> Instance for Mapping<K, V> {
     }
 }
 
+#[cfg(feature = "mock-vm")]
```
Why Mapping needs serialization?

kpob replied:
Modules (including Mapping and Variable) are no longer serialized

zie1ony commented on examples/src/erc20.rs:
```diff
@@ -12,7 +12,7 @@ pub struct Erc20 {
     total_supply: Variable<U256>,
     balances: Mapping<Address, U256>,
-    allowances: Mapping<(Address, Address), U256>
+    allowances: Mapping<Address, Mapping<Address, U256>>
```
I'd keep the old style. It's probably faster. Noo need to to create intermediate Mapping here. Setters and getters are more readable.

zie1ony commented on examples/src/mapping.rs:
```diff
@@ -0,0 +1,229 @@
+impl NestedMapping {
+    pub fn set_string(&mut self, key1: String, key2: u32, key3: String, value: String) {
+        map!(self.strings[key1][key2][key3] = value.clone());
+        map!(self.tokens[key1][key2][key3] = OwnedToken::instance(&value));
```
Is `value` a namespace?
Do we want developers to set `instance` manually? I think not. It should be constructed only using getter on `Mapping`.

zie1ony commented on examples/src/mapping.rs:
```diff
@@ -0,0 +1,229 @@
+    }
+
+    pub fn balance_of(&self, token_name: String, owner: Address) -> U256 {
+        let token = self.get_token(token_name);
```
Could be one line.

zie1ony commented on examples/src/mapping.rs:
```diff
@@ -0,0 +1,229 @@
+    }
+
+    pub fn mint(&mut self, token_name: String, account: Address, amount: U256) {
+        let mut token = self.get_token(token_name);
```
Could be one line.

zie1ony commented on core/src/variable.rs:
```diff
@@ -81,6 +81,54 @@ impl<T: OdraType> Instance for Variable<T> {
     }
 }
 
+#[cfg(feature = "mock-vm")]
```
Same question as in mapping. Why need to serialize?

zie1ony commented on lang/ir/src/mapping.rs:
```diff
@@ -0,0 +1,123 @@
+use proc_macro2::{TokenStream, TokenTree};
```
Nice implementation :)

zie1ony commented on odra-casper/backend/src/casper_env.rs:
```diff
@@ -130,12 +130,24 @@ where
 }
 
+/// Convert any key to hash.
+fn to_dictionary_key<T: ToBytes>(seed: &str, key: &T) -> String {
```
I think it makes sense to cache the result of hashing.

zie1ony commented on odra-casper/backend/src/casper_env.rs:
```diff
@@ -217,3 +229,41 @@ fn is_purse_empty(purse: URef) -> bool {
         .unwrap_or_else(|| true)
 }
+
+fn get_new_dict_uref() -> URef {
```
Why not using `casper_contract::contract_api::storage::new_dictionary` and then `runtime::remove_key`? Less unsafe code is better.

kpob replied:
Yup, replaced with casper api, and it is much cleaner now.

zie1ony commented on odra-casper/backend/src/casper_env.rs:
```diff
@@ -217,3 +229,41 @@ fn is_purse_empty(purse: URef) -> bool {
+    bytesrepr::deserialize(dest).unwrap_or_revert()
+}
+
+fn get_state_uref() -> URef {
```
This should definitely be cached.

kpob replied:
Done

zie1ony commented on lang/codegen/src/generator/mapping.rs:
```diff
@@ -0,0 +1,61 @@
+                .take(segments_count - 1)
+                .map(|e| quote!(let v = v.get_instance(&#e);))
+                .collect::<TokenStream>();
+            let return_value = quote!(odra::UnwrapOrRevert::unwrap_or_revert(v.get(&#value)));
```
Instance should always be there. Is it possible to write it without `unwrap_or_revert`?

kpob replied:
Because we removed modules serialization, and modules are no longer of `OdraType`, so we can't apply map!(a[b][c]) syntax to a module. But if the value is of `OdraType` eg. `String` then we call `get()` and uwrap the optional value (no guarantees the types has a default value).

zie1ony commented on lang/codegen/src/generator/mapping.rs:
```diff
@@ -0,0 +1,61 @@
+            let value_assign = quote!(v.set(&#value, #assigned_value));
+
+            quote! {
+                #init_root
```
Shouldn't this also be wrapped with brackets `{ }`?

kpob replied:
No, setters don't return a value, so a getter must be inside a block but a setter doesn't

kpob committed:
Improve public methods and serialization for modules

- Update trait bounds for `Mapping` struct, new and get methods
- Add `Mapping` `get_instance` and `From` implementations
- Update entrypoints to use public methods instead of all methods
- Rename `get_method_iter` to `get_public_method_iter`
- Remove dependency on `common` module for serialization
- Add `Instance` derive for instantiable modules
- Remove mock and casper serialization

[core/src/mapping.rs]
- Change the trait bound for `V` in the `Mapping` struct
- Change the trait bound for `V` in the `Mapping` `new` method
- Change the trait bound for `V` in the `Mapping` `get` method
- Add a `Mapping` `get_instance` method
- Add a `Mapping` `From` implementation
- Add a `Mapping` `Instance` implementation
[lang/codegen/src/generator/module_impl/deploy.rs]
- Change the entrypoints to use public methods instead of all methods
[lang/codegen/src/generator/module_impl/reference.rs]
- Update entrypoints to use public methods instead of all methods
[lang/ir/src/module_item/module_impl.rs]
- Rename `get_method_iter` to `get_public_method_iter`
- Add a filter to only return public methods for trait implementations
[lang/codegen/src/generator/module_item.rs]
- Remove dependency on `common` module for serialization
- Add `Instance` derive for instantiable modules
- Remove mock and casper serialization
[examples/src/mapping.rs]
- Add `UnwrapOrRevert` import
- Change `Instance` to `UnwrapOrRevert` in `set_string`
- Remove cloning of value in `set_string`
- Simplify `init` call in `set_string`
- Add `UnwrapOrRevert` in `get_string_api`
- Simplify `total_supply` call
- Add `init` call in `add_token

kpob committed:
Add macro and boilerplate for nested map access

- Add a macro for accessing values in a nested map
- Implement boilerplate code for reading/writing values from/to a map

[lang/proc-macros/src/lib.rs]
- Add a macro for accessing values in a nested map
- Implement boilerplate code for reading/writing values from/to a map

kpob committed:
- Replace `get_new_dict_uref()` with `storage::new_dictionary()`
- Cache state URef

kpob committed:
Increase performance of Casper environment variables

- Add `KEYS` static variable to `casper_env.rs`
- Update `to_dictionary_key` function to cache hash results

zie1ony reviewed (approved):
Great :)

kpob committed:
Move TokenManager example to a new module
